quickjs.c

```diff
@@ -51507,3 +51631,166 @@ uint32_t add_ic_slot1(JSContext *ctx, JSInlineCache *ic, JSAtom atom)
+
+    JS_SetOpaque(obj, csd1);
+
+    return obj;
+}
+
+static void js_new_callsite_data(JSContext *ctx, JSCallSiteData *csd, JSStackFrame *sf)
+{
+    const char *func_name_str;
+    JSObject *p;
+
+    csd->func = js_dup(sf->cur_func);
+    func_name_str = get_func_name(ctx, sf->cur_func);
+    if (!func_name_str || func_name_str[0] == '\0')
+        csd->func_name = JS_NULL;
+    else
+        csd->func_name = JS_NewString(ctx, func_name_str);
+    JS_FreeCString(ctx, func_name_str);
+    if (JS_IsException(csd->func_name))
+        csd->func_name = JS_NULL;
+
+    p = JS_VALUE_GET_OBJ(sf->cur_func);
+    if (js_class_has_bytecode(p->class_id)) {
+        JSFunctionBytecode *b = p->u.func.function_bytecode;
+        int line_num1, col_num1;
+        line_num1 = find_line_num(ctx, b,
+                                  sf->cur_pc - b->byte_code_buf - 1,
+                                  &col_num1);
+        csd->native = FALSE;
+        csd->line_num = line_num1;
+        csd->col_num = col_num1;
+        csd->filename = JS_AtomToString(ctx, b->filename);
+    } else {
+        csd->native = TRUE;
+        csd->line_num = -1;
+        csd->col_num = -1;
+        csd->filename = JS_NULL;
+    }
+}
+
+static void js_new_callsite_data2(JSContext *ctx, JSCallSiteData *csd, const char *filename, int line_num, int col_num)
+{
+    csd->func = JS_NULL;
+    csd->func_name = JS_NULL;
+    csd->native = FALSE;
+    csd->line_num = line_num;
+    csd->col_num = col_num;
+    csd->filename = JS_NewString(ctx, filename);
+    if (JS_IsException(csd->filename))
+        csd->filename = JS_NULL;
+}
+
+static JSValue js_callsite_getcolumnnumber(JSContext *ctx, JSValue this_val, int argc, JSValue *argv)
+{
+    JSCallSiteData *csd = JS_GetOpaque2(ctx, this_val, JS_CLASS_CALL_SITE);
+    if (!csd)
+        return JS_EXCEPTION;
+    return JS_NewInt32(ctx, csd->col_num);
+}
+
+static JSValue js_callsite_getfilename(JSContext *ctx, JSValue this_val, int argc, JSValue *argv)
+{
+    JSCallSiteData *csd = JS_GetOpaque2(ctx, this_val, JS_CLASS_CALL_SITE);
+    if (!csd)
+        return JS_EXCEPTION;
+    return js_dup(csd->filename);
+}
+
+static JSValue js_callsite_getfunction(JSContext *ctx, JSValue this_val, int argc, JSValue *argv)
+{
+        JSCallSiteData *csd = JS_GetOpaque2(ctx, this_val, JS_CLASS_CALL_SITE);
+    if (!csd)
+        return JS_EXCEPTION;
+    return js_dup(csd->func);
+}
+
+static JSValue js_callsite_getfunctionname(JSContext *ctx, JSValue this_val, int argc, JSValue *argv)
+{
+    JSCallSiteData *csd = JS_GetOpaque2(ctx, this_val, JS_CLASS_CALL_SITE);
+    if (!csd)
+        return JS_EXCEPTION;
+    return js_dup(csd->func_name);
+}
+
+static JSValue js_callsite_getlinenumber(JSContext *ctx, JSValue this_val, int argc, JSValue *argv)
+{
+    JSCallSiteData *csd = JS_GetOpaque2(ctx, this_val, JS_CLASS_CALL_SITE);
+    if (!csd)
+        return JS_EXCEPTION;
+    return JS_NewInt32(ctx, csd->line_num);
+}
```

Comment:
Just a suggestion but you can DRY these using JS_CFUNC_DEF with `magic == offsetof(JSCallSiteData, field)` and then compute the address with `int *field = (void *)((char *)csd + magic);`